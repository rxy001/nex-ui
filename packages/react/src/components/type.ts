export type ClassNames = string | string[]

export type ComponentClasses<OwnerState, T extends string> = Partial<
  Record<T, ClassNames | ((ownerState: OwnerState) => ClassNames)>
>
