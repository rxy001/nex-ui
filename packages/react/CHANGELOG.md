# @nex-ui/react

## 0.0.15

### Patch Changes

- fdd4483: Release
- Updated dependencies [fdd4483]
  - @nex-ui/styled@0.0.15
  - @nex-ui/system@0.0.15
  - @nex-ui/hooks@0.0.15
  - @nex-ui/icons@0.0.15
  - @nex-ui/utils@0.0.15

## 0.0.14

### Patch Changes

- ca25983: Release
- Updated dependencies [ca25983]
  - @nex-ui/styled@0.0.14
  - @nex-ui/system@0.0.14
  - @nex-ui/hooks@0.0.14
  - @nex-ui/icons@0.0.14
  - @nex-ui/utils@0.0.14

## 0.0.13

### Patch Changes

- 1b97788: Release
- Updated dependencies [1b97788]
  - @nex-ui/styled@0.0.13
  - @nex-ui/system@0.0.13
  - @nex-ui/hooks@0.0.13
  - @nex-ui/icons@0.0.13
  - @nex-ui/utils@0.0.13

## 0.0.12

### Patch Changes

- 664eb21: Test

## 0.0.11

### Patch Changes

- 259356e: Test
- Updated dependencies [259356e]
  - @nex-ui/styled@0.0.11
  - @nex-ui/system@0.0.11
  - @nex-ui/hooks@0.0.11
  - @nex-ui/icons@0.0.11
  - @nex-ui/utils@0.0.10

## 0.0.10

### Patch Changes

- 5a229bd: release
- Updated dependencies [5a229bd]
  - @nex-ui/styled@0.0.10
  - @nex-ui/system@0.0.10
  - @nex-ui/icons@0.0.10

## 0.0.9

### Patch Changes

- 4a5c359: release
- Updated dependencies [4a5c359]
  - @nex-ui/styled@0.0.9
  - @nex-ui/system@0.0.9
  - @nex-ui/hooks@0.0.9
  - @nex-ui/icons@0.0.9
  - @nex-ui/utils@0.0.9

## 0.0.8

### Patch Changes

- c8c1956: release
- Updated dependencies [c8c1956]
  - @nex-ui/styled@0.0.8
  - @nex-ui/system@0.0.8
  - @nex-ui/hooks@0.0.8
  - @nex-ui/icons@0.0.8
  - @nex-ui/utils@0.0.8

## 0.0.7

### Patch Changes

- 2618e1a: release
- Updated dependencies [2618e1a]
  - @nex-ui/styled@0.0.7
  - @nex-ui/system@0.0.7
  - @nex-ui/hooks@0.0.7
  - @nex-ui/icons@0.0.7
  - @nex-ui/utils@0.0.7

## 0.0.6

### Patch Changes

- 87191c1: release
- Updated dependencies [87191c1]
  - @nex-ui/styled@0.0.6
  - @nex-ui/system@0.0.6
  - @nex-ui/hooks@0.0.6
  - @nex-ui/icons@0.0.6
  - @nex-ui/utils@0.0.6

## 0.0.5

### Patch Changes

- 7631577: release
- Updated dependencies [7631577]
  - @nex-ui/styled@0.0.5
  - @nex-ui/system@0.0.5
  - @nex-ui/hooks@0.0.5
  - @nex-ui/icons@0.0.5
  - @nex-ui/utils@0.0.5

## 0.0.4

### Patch Changes

- 70cb478: v0.0.4
- Updated dependencies [70cb478]
  - @nex-ui/styled@0.0.4
  - @nex-ui/system@0.0.4
  - @nex-ui/hooks@0.0.4
  - @nex-ui/icons@0.0.4
  - @nex-ui/utils@0.0.4
