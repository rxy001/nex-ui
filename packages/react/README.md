<div style="background:white;">
  <p align="center">
    <img src='../../assets/logo.png' style="width: 150px; height: 150px;" />
  </p>
  <h1 align="center">Nex UI</h1>
  <p align="center">ðŸŽ‰ A beautiful, modern, and reliable React component library.</p>
  <p align="center"><b>Theme Customizable, Light & Dark UI, Unique DX, Written In TypeScript</b></p>
</div>

## Documentation

Visit https://nex-ui-docs.vercel.app to view the full documentation.

## Getting Started

Visit https://nex-ui-docs.vercel.app/docs/getting-started to get started with Nex UI.

## Storybook

Visit https://nex-ui-storybook.vercel.app to view the storybook for all components.

## Contributing

Contributions are always welcome!

See [CONTRIBUTING.md](/CONTRIBUTING.md) for ways to get started.

Please adhere to this project's [CODE_OF_CONDUCT](/CODE_OF_CONDUCT.md).

## License

[MIT](https://choosealicense.com/licenses/mit/)
