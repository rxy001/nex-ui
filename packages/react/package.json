{
  "name": "@nex-ui/react",
  "version": "0.0.12",
  "author": "X1ng Yu",
  "description": "ðŸŽ‰ A beautiful, modern, and reliable React component library.",
  "homepage": "https://nex-ui-docs.vercel.app/",
  "private": false,
  "keywords": [
    "react",
    "react-component",
    "nui",
    "nex-ui"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rxy001/nex-ui.git",
    "directory": "packages/react"
  },
  "license": "MIT",
  "scripts": {
    "build": "tsx ../../scripts/build.ts --dts",
    "build:fast": "tsx ../../scripts/build.ts",
    "typecheck": "tsc --noEmit",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "type": "module",
  "files": [
    "dist"
  ],
  "main": "src/index.ts",
  "devDependencies": {
    "@storybook/react": "8.6.12"
  },
  "dependencies": {
    "@nex-ui/hooks": "workspace:*",
    "@nex-ui/icons": "workspace:*",
    "@nex-ui/styled": "workspace:*",
    "@nex-ui/system": "workspace:*",
    "@nex-ui/utils": "workspace:*",
    "clsx": "2.1.1"
  },
  "peerDependencies": {
    "@emotion/react": ">=11",
    "motion": "^12.7.2",
    "react": ">=19",
    "react-dom": ">=19"
  },
  "sideEffects": false,
  "clean-package": "../../clean-package.config.json"
}
