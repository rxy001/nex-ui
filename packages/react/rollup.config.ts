import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { defineConfig } from 'rollup'
import typescript from '@rollup/plugin-typescript'
import del from 'rollup-plugin-delete'
import dts from 'rollup-plugin-dts'
import { nexUIRollupPlugin } from '@nex-ui/plugins'

const dirname = fileURLToPath(new URL('./', import.meta.url))

export default defineConfig([
  {
    plugins: [
      del({
        targets: './dist/*',
      }),
      typescript({
        tsconfig: path.resolve(dirname, 'tsconfig.json'),
      }),
      nexUIRollupPlugin(),
    ],
    input: './src/index.ts',
    external: [
      'react',
      'react-dom',
      'react/jsx-runtime',
      'react-dom/client',
      '@nex-ui/utils',
      '@nex-ui/css-system',
      'classnames',
      'framer-motion',
      '@iconify/react',
      '@vanilla-extract/dynamic',
      '@vanilla-extract/css',
      '@vanilla-extract/css/functionSerializer',
    ],
    output: [
      {
        preserveModules: true,
        format: 'es',
        dir: './dist/es',
        assetFileNames({ name }) {
          return name?.replace(/^src\//, '') ?? ''
        },
      },
      {
        preserveModules: true,
        format: 'cjs',
        dir: './dist/lib',
        assetFileNames({ name }) {
          return name?.replace(/^src\//, '') ?? ''
        },
      },
    ],
  },
  {
    plugins: [dts()],
    input: './src/index.ts',
    output: {
      dir: './dist/types',
      format: 'es',
      preserveModules: true,
    },
  },
])
