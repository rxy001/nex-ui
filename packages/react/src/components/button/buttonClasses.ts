import { generateUtilityClasses } from '../utils'

// 用于单元测试
export const buttonClasses = generateUtilityClasses('nui-btn', [
  // ---------root---------
  'root',
  // "link" | "text" | "solid" | "outline"
  'variant-link',
  'variant-text',
  'variant-solid',
  'variant-outline',

  // "sm" | "md" | "lg" | "full"
  'radius-sm',
  'radius-md',
  'radius-lg',
  'radius-full',

  // "blue" | "gray" | "pink" | "purple" | "cyan" | "yellow" | "orange" | "red" | "green"
  'color-gray',
  'color-green',
  'color-red',
  'color-yellow',
  'color-blue',
  'color-pink',
  'color-cyan',
  'color-purple',
  'color-orange',

  // "sm" | "md" | "lg"
  'size-sm',
  'size-md',
  'size-lg',

  'icon-only',
  'loading',
  'disabled',
  'block',
  // ---------root---------

  'icon',
  'start-icon',
  'end-icon',

  'icon-size-sm',
  'icon-size-md',
  'icon-size-lg',

  'icon-loading',
])
