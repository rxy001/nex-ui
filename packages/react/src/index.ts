import './types/declare'

export * from './components'
export * from './theme'

export type { Theme } from './types/theme'
export type { ComponentsTheme } from './types/componentsTheme'

export { useColorScheme } from '@nex-ui/system'
export { nex, styled } from '@nex-ui/styled'
