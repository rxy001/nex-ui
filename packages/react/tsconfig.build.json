{
  "extends": "../../tsconfig.json",
  "include": ["./src"],
  "exclude": ["./src/**/__tests__/**/*", "./src/**/__stories__/**/*"]
}
