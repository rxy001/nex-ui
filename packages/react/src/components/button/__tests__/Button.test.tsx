import { useState, createRef } from 'react'
import { fireEvent, act } from '@testing-library/react'
import {
  testColorClasses,
  testComponentStability,
  renderWithNexUIProvider,
  testRootClassName,
  testVariantClasses,
  testRadiusClasses,
} from '~/tests/shared'
import { Button } from '../index'
import { buttonClasses } from '../buttonClasses'
import type { ButtonProps } from '../index'

describe('Button', () => {
  testComponentStability(<Button />)

  testRootClassName(<Button className='test-class' />, 'test-class')

  testColorClasses(<Button>Button</Button>, buttonClasses)

  testVariantClasses(
    <Button>Button</Button>,
    ['variant', ['solid', 'outlined', 'text']],
    buttonClasses,
  )

  testRadiusClasses(<Button>Button</Button>, buttonClasses)

  it('should render with default props', () => {
    const { container } = renderWithNexUIProvider(<Button>Button</Button>)
    const button = container.firstElementChild

    expect(button).toHaveClass(buttonClasses.root)
    expect(button).toHaveClass(buttonClasses['variant-solid'])
    expect(button).toHaveClass(buttonClasses['size-md'])
    expect(button).toHaveClass(buttonClasses['radius-md'])
    expect(button).toHaveClass(buttonClasses['color-blue'])

    expect(button).not.toHaveClass(buttonClasses['variant-outlined'])
    expect(button).not.toHaveClass(buttonClasses['variant-text'])
    expect(button).not.toHaveClass(buttonClasses['color-green'])
    expect(button).not.toHaveClass(buttonClasses['color-cyan'])
    expect(button).not.toHaveClass(buttonClasses['color-orange'])
    expect(button).not.toHaveClass(buttonClasses['color-pink'])
    expect(button).not.toHaveClass(buttonClasses['color-purple'])
    expect(button).not.toHaveClass(buttonClasses['color-yellow'])
    expect(button).not.toHaveClass(buttonClasses['color-red'])
    expect(button).not.toHaveClass(buttonClasses['color-gray'])
    expect(button).not.toHaveClass(buttonClasses['size-sm'])
    expect(button).not.toHaveClass(buttonClasses['size-lg'])
    expect(button).not.toHaveClass(buttonClasses['radius-sm'])
    expect(button).not.toHaveClass(buttonClasses['radius-lg'])
    expect(button).not.toHaveClass(buttonClasses['icon-only'])
    expect(button).not.toHaveClass(buttonClasses.loading)
    expect(button).not.toHaveClass(buttonClasses.disabled)
    expect(button).not.toHaveClass(buttonClasses['full-width'])
    expect(button).not.toHaveClass(buttonClasses['disable-ripple'])

    expect(button).toMatchSnapshot()
  })

  it("should forward ref to Button's root element", () => {
    const ref = createRef<HTMLButtonElement>()
    const { container } = renderWithNexUIProvider(
      <Button ref={ref}>Button</Button>,
    )
    const button = container.firstElementChild as HTMLButtonElement
    expect(ref.current).toBe(button)
  })

  it('should add the appropriate iconOnly class to root element based on iconOnly prop', () => {
    const { getByTestId, rerender } = renderWithNexUIProvider(
      <Button iconOnly data-testid='icon-only'>
        Button
      </Button>,
    )

    expect(getByTestId('icon-only')).toHaveClass(buttonClasses['icon-only'])

    rerender(<Button data-testid='not-icon-only'>Button</Button>)
    expect(getByTestId('not-icon-only')).not.toHaveClass(
      buttonClasses['icon-only'],
    )
  })

  it('should add the appropriate fullWidth class to root element based on fullWidth prop', () => {
    const { getByTestId, rerender } = renderWithNexUIProvider(
      <Button fullWidth data-testid='full-width'>
        Button
      </Button>,
    )

    expect(getByTestId('full-width')).toHaveClass(buttonClasses['full-width'])

    rerender(<Button data-testid='not-full-width'>Button</Button>)
    expect(getByTestId('not-full-width')).not.toHaveClass(
      buttonClasses['full-width'],
    )
  })

  it('should add the appropriate disableRipple class to root element based on disableRipple prop', () => {
    const { getByTestId, rerender } = renderWithNexUIProvider(
      <Button disableRipple data-testid='disable-ripple'>
        Button
      </Button>,
    )

    expect(getByTestId('disable-ripple')).toHaveClass(
      buttonClasses['disable-ripple'],
    )

    rerender(<Button data-testid='not-disable-ripple'>Button</Button>)
    expect(getByTestId('not-disable-ripple')).not.toHaveClass(
      buttonClasses['disable-ripple'],
    )
  })

  it('should trigger onClick function', async () => {
    const onClick = jest.fn()
    const { getByRole } = renderWithNexUIProvider(<Button onClick={onClick} />)
    await act(async () => {
      fireEvent.click(getByRole('button'))
    })
    expect(onClick).toHaveBeenCalled()
  })

  it('should have disabled class when disabled', async () => {
    const onClick = jest.fn()
    const { getByText } = renderWithNexUIProvider(
      <Button disabled onClick={onClick}>
        Button
      </Button>,
    )

    const button = getByText('Button')
    expect(button).toHaveClass(buttonClasses.disabled)
  })

  it('should disable the button when loading', () => {
    const { getByTestId } = renderWithNexUIProvider(
      <Button loading data-testid='loading-button'>
        Button
      </Button>,
    )

    const button = getByTestId('loading-button')
    expect(button).toHaveClass(buttonClasses.loading)
    expect(button).toHaveStyleRule('pointer-events', 'none')
  })

  it('should support link button', () => {
    const { container } = renderWithNexUIProvider(
      <Button target='_blank' href='https://'>
        Link Button
      </Button>,
    )
    expect(container.firstElementChild).toMatchSnapshot()
  })

  it('should render with start icon', () => {
    const { getByTestId } = renderWithNexUIProvider(
      <Button startIcon={<span data-testid='start-icon'>Icon</span>}>
        Button
      </Button>,
    )
    const startIcon = getByTestId('start-icon')
    expect(startIcon).toBeInTheDocument()
    expect(startIcon.parentElement).toHaveClass(buttonClasses.icon)
    expect(startIcon.parentElement).toHaveClass(buttonClasses['start-icon'])
  })

  it('should render with end icon', () => {
    const { getByTestId } = renderWithNexUIProvider(
      <Button endIcon={<span data-testid='end-icon'>Icon</span>}>
        Button
      </Button>,
    )
    const endIcon = getByTestId('end-icon')
    expect(endIcon).toBeInTheDocument()
    expect(endIcon.parentElement).toHaveClass(buttonClasses.icon)
    expect(endIcon.parentElement).toHaveClass(buttonClasses['end-icon'])
  })

  it('should support to change loading', async () => {
    const DefaultButton: React.FC = () => {
      const [loading, setLoading] = useState<ButtonProps['loading']>(false)
      return (
        <Button loading={loading} onClick={() => setLoading(true)}>
          Button
        </Button>
      )
    }
    const { getByRole } = renderWithNexUIProvider(<DefaultButton />)
    const button = getByRole('button')

    await act(async () => {
      fireEvent.click(button)
    })
    expect(button).toHaveClass(buttonClasses.loading)
    expect(
      button.querySelector(`.${buttonClasses['icon-loading']}`),
    ).toBeInTheDocument()
  })

  it('should forward classes to root, startIcon and endIcon slots', () => {
    const classes = {
      startIcon: 'test-start-icon-class',
      endIcon: 'test-end-icon-class',
    }

    const { container } = renderWithNexUIProvider(
      <Button
        classes={classes}
        startIcon={<span>start icon</span>}
        endIcon={<span>end icon</span>}
      >
        Button
      </Button>,
    )

    const button = container.firstElementChild
    const startIcon = button?.querySelector(`.${buttonClasses['start-icon']}`)
    const endIcon = button?.querySelector(`.${buttonClasses['end-icon']}`)

    expect(button).toHaveClass(buttonClasses.root)
    expect(startIcon).toHaveClass(classes.startIcon)
    expect(endIcon).toHaveClass(classes.endIcon)
  })

  it('should support customized spinner', () => {
    const { container } = renderWithNexUIProvider(
      <Button
        loading
        spinner={<span data-testid='custom-spinner'>Custom Spinner</span>}
      >
        Button
      </Button>,
    )

    const button = container.firstElementChild
    const spinner = button?.querySelector(`.${buttonClasses['icon-loading']}`)

    expect(spinner).toBeInTheDocument()
    expect(spinner).toHaveTextContent('Custom Spinner')
  })

  it('should support spinner placement', () => {
    const { container, rerender } = renderWithNexUIProvider(
      <Button loading spinnerPlacement='start'>
        Button
      </Button>,
    )

    const startSpinner = container.firstElementChild?.querySelector(
      `.${buttonClasses['start-icon']}`,
    )

    expect(startSpinner).toBeInTheDocument()

    rerender(
      <Button loading spinnerPlacement='end'>
        Button
      </Button>,
    )

    const endSpinner = container.firstElementChild?.querySelector(
      `.${buttonClasses['end-icon']}`,
    )
    expect(endSpinner).toBeInTheDocument()
  })

  it('should forward slotProps to startIcon and endIcon slots', () => {
    const { container } = renderWithNexUIProvider(
      <Button
        startIcon={<span data-testid='start-icon'>Start Icon</span>}
        endIcon={<span data-testid='end-icon'>End Icon</span>}
        slotProps={{
          startIcon: { className: 'test-start-icon' },
          endIcon: { className: 'test-end-icon' },
        }}
      >
        Button
      </Button>,
    )

    const startIcon = container.querySelector('.test-start-icon')
    const endIcon = container.querySelector('.test-end-icon')

    expect(startIcon).toBeInTheDocument()
    expect(endIcon).toBeInTheDocument()
  })
})
