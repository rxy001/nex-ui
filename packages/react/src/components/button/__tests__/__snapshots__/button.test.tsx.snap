// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<button
  class="nui-btn nui-btn-variant-solid nui-btn-radius-md nui-btn-size-md nui-btn-color-blue emotion-0"
  type="button"
>
  Button
</button>
`;

exports[`Button should support link button 1`] = `
<a
  class="nui-btn nui-btn-variant-solid nui-btn-radius-md nui-btn-size-md nui-btn-color-blue emotion-0"
  href="https://"
  target="_blank"
>
  Link Button
</a>
`;
